### PR TITLE
Add Intel compiler to settings and manage compatibility with base compiler

### LINKED ISSUE #5590: Add intel compiler to default settings.yml
Suggested:

```
intel:
        version: ["17.0.0", "17.0.1", "17.0.2", "17.0.3", "17.0.4", "17.0.5", "17.0.6", "17.0.7", "17.0.8",
                  "18.0.0", "18.0.1", "18.0.2", "18.0.3", "18.0.4",
                  "19.0.0", "19.0.1", "19.0.2", "19.0.3"]
        libcxx: [libstdc++, libstdc++11] # Linux only
        runtime: [MD, MT, MTd, MDd] # Windows only
        cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20]
```

To investigate:
- Binary compatibility of update/patches: reported erorrs on 17.0.4->17.0.6
- Intel compiler also includes fortran, check if it makes sense regarding cppstd/libcxx
- libcxx/runtime in different OS, how to manage?

### PR DESCRIPTION
**Pull request by danimtb**
Changelog: Feature: Add intel compiler to *settings.yml* and manage compatibility
Docs: missing

- [x] Refer to the issue that supports this Pull Request: closes #5590 
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/model/info.py [hunk 1]
```diff
@@ -475,3 +476,42 @@ def default_std_non_matching(self):
...
+            # Set base_incompatible=True in order to make packages compatible
+            self.settings.compiler.base_incompatible = True
+
+            # Deep copy everything
+            for field, value in self.full_settings.compiler.as_list():
+                tokens = field.split(".")
+                attr = self.settings.compiler
+                for token in tokens[:-1]:
+                    attr = getattr(attr, token)
+                setattr(attr, tokens[-1], value)
```
Comment (ohanar) [inline conans/model/info.py]: Both of these methods currently reset any package_id manipulation already done to the compiler. Should we try to preserve previous changes?

**Comment by ohanar** [comment]
Do you want to include the build helper changes here? If so, I can make a pull request to your pull request with the changes that I have locally, otherwise, I'll open another pull request alongside this one.

**Comment by ohanar** [comment]
One thing that should be addressed is how to handle minor intel compiler versions. Sometimes intel adds pretty significant functionality to these, even though they are supposedly minor updates (e.g. 19.0.4 added official support for Visual Studio 2019). Should we do the same thing as gcc, i.e. having 19.0 as well as 19.0.0, 19.0.1, etc?

**Comment by ohanar** [comment]
I've pushed some initial support for the intel compiler in the build helpers [here](https://github.com/ohanar/conan/tree/intel_compiler).

File: conans/model/info.py (hunk 1 above)
Comment (danimtb) [reply conans/model/info.py]: yes, maybe we have to preserve those

**Comment by danimtb** [comment]
Hi @ohanar and thanks a lot for your comments. I think we will add the changes for the build helper in a different PR.

Regarding the versioning of the Intel, I don't know how we should approach it cause there big features coming out in patch releases... Is there any statement of compatibility in versions from Intel?

File: conans/model/info.py (hunk 1 above)
Comment (danimtb) [reply conans/model/info.py]: I am having a look at the other methods and those seem to be implemented in the same way. So I think I will leave this like this for now

**Comment by danimtb** [comment]
@ohanar would you like to create the PR for the cmake helper yourself or would you like me to do it for you?

I would like to have those changes independent from this PR so it is easier to review. Thanks!

**Comment by ohanar** [comment]
There is this note from intel about binary compatibly between versions of their compiler is [here](https://software.intel.com/en-us/articles/are-intel-compiler-generated-objectsbinaries-backward-andor-forward-compatible-between). This only addresses major versions, but it seems reasonable to assume it applies to minor versions as well. It seems like objects are generally forward compatible across versions, so long as the "base" api is the same and ipo is not used (which pretty much every compiler says). I think it is probably okay to mimic gcc usage here, that is prefer the major version, but allow users who  really need to distinguish binaries produced by different minor version (e.g. for static libraries with ipo).

As for the pull request, I'm happy to open one, however I'll be on vacation next week so I won't be around to address feedback, or finish functionality (tests, old compiler versions). So it might be better if you were to open a pull request.

**Comment by danimtb** [comment]
Thanks a lot for the link! I will create a new PR with your code for the CMake build helper and ping you there so you can review it whenever possible 😄

File: conans/client/conf/__init__.py [hunk 2]
```diff
@@ -86,6 +86,17 @@
         version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0"]
         libcxx: [libstdc++, libc++]
         cppstd: [None, 98, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20]
+    intel:
+        version: ["11", "12", "13", "14", "15", "16", "16.1", "17", "18", "19"]
```
[the remark this answers is not in the document]
Comment (danimtb) [reply conans/client/conf/__init__.py]: yeah, my fault. Thanks for taking a look

File: conans/test/integration/package_id_test.py [hunk 3]
```diff
@@ -430,3 +430,78 @@ def test_std_matching_with_compiler_cppstd(self):
...
+
+        self.client.run('install Hello/1.2.0@user/testing '
+                        '-s compiler="intel" '
+                        '-s compiler.version="16" '
+                        '-s compiler.base_incompatible=True '
+                        '-s compiler.base="Visual Studio" '
+                        '-s compiler.base.version=8 '
+                        '-s compiler.base.runtime=MD')
+        self.assertIn("Hello/1.2.0@user/testing:1151fe341e6b310f7645a76b4d3d524342835acc - Cache",
+                      self.client.out)
```
[the remark this answers is not in the document]
Comment (danimtb) [reply conans/test/integration/package_id_test.py]: yes, that is the compatibility behavior we are looking for. However, we now realized that we need to take the intel runtime into account too, so this model has to be completed.

**Comment by lasote** [comment]
Follow here: https://github.com/conan-io/conan/pull/6052

**Pull request closed without merging**